lib/src/log.rs

```diff
@@ -129,22 +130,84 @@ pub fn reset_task_log_directory(path: &Path) -> Result<(), Error> {
 
+pub fn read_last_lines_as_byte_deque(file: &mut File, lines: usize) -> Result<VecDeque<u8>, Error> {
+    // The integer types were chosen to avoid overflows when converting between signed and unsigned
+    // integers. u16 is used here so that we can support 16bit platforms with simple integer casts
+    // (FIXME Not sure if pueue even is supported on any 16bit platforms)
+    let chunk_size = 4096u16;
+    let mut buf = vec![0u8; chunk_size as usize];
+    let size = match file.seek(io::SeekFrom::End(0)) {
+        Ok(p) => p,
+        Err(err) => {
+            return Err(Error::LogSeek(format!(
+                "Error while seeking log file : {}",
+                err
+            )));
+        }
+    };
+    let mut position = size;
+    let mut text = std::collections::VecDeque::new();
+    let mut lines_seen = 0usize;
+    'outer: loop {
+        if position == 0 {
+            break;
+        }
+        // Safe to cast chunk_size from u16 to u64
+        let next_chunk_size = if position > chunk_size as u64 {
+            chunk_size
+        } else {
+            // Safe to cast position to u16 since it is <= chunk_size which is u16
+            position as u16
+        };
+        // Safe to cast next_chunk_size from u16 to i64
+        position = match file.seek(io::SeekFrom::Current(-(next_chunk_size as i64))) {
+            Ok(p) => p,
+            Err(err) => {
+                return Err(Error::LogSeek(format!(
+                    "Error while seeking log file : {}",
+                    err
+                )));
+            }
+        };
+        // Safe to cast next_chunk_size from u16 to usize
+        let slice = &mut buf[..next_chunk_size as usize];
+        if let Err(err) = file.read_exact(slice) {
+            return Err(Error::LogRead(format!(
+                "Error while reading log file : {}",
+                err
+            )));
+        }
+        for b in slice.iter().rev() {
+            if *b == b'\n' {
+                lines_seen += 1;
+                if lines_seen == lines + 1 {
+                    break 'outer;
+                }
+            }
+            text.push_front(*b)
+        }
+    }
+    if let Err(err) = file.seek(io::SeekFrom::End(0)) {
+        return Err(Error::LogSeek(format!(
+            "Error while seeking log file : {}",
+            err
+        )));
+    };
+    Ok(text)
+}
+
 /// Read the last `amount` lines of a file to a string.
-///
-/// TODO: This is super imperformant, but works as long as we don't use the last
-/// 1000 lines. It would be cleaner to seek to the beginning of the requested
-/// position and simply stream the content.
 pub fn read_last_lines(file: &mut File, amount: usize) -> String {
-    // Read the lines from the file.
-    // Return a debug string if it fails.
-    let last_lines: Vec<String> = match rev_lines::RevLines::new(BufReader::new(file)) {
-        Ok(lines) => lines.take(amount).collect(),
+    match read_last_lines_as_byte_deque(file, amount) {
```

Comment:
After reading through my old docs and the old implementation, I'm not 100% sure if your current implementation changes anything at all.
`rev_lines` also read the file from the back in a BufReader style, but with a small convenience layer on top.

The part about imperformant code mentioned in the docstring isn't about the bufreader implementation, but rather about reading all lines into memory, reversing them and only then streaming them to disk.

This gets troublesome, if users try to read a few thousand lines with possibly gigabytes of data from the logs. This already happened and lead to RAM exhaustion.

A proper performant solution would be to read through the file until X+1 `\n` are found.
The character after the `\n` is then the start of our log output and we could simply pass the handle to a io::copy call with stdout as the sink.
`\n\r` should obviously also be handled

This function would then be changed to something like this:
```
/// Seek to the beginning of a line, which is located `lines` newlines (`\n`) from the end of the file.
/// Jumps to the top of the files, if it doesn't contain enough files.
/// Returns the number of lines that were found.
pub fn seek_to_line_rev(file: &mut File, lines: usize) -> Result<usize> {}
```

Reply:
FWIW, I believe my implementation does in fact read in 4k chunks going backwards. I agree generally it could use more documentation, but I sort of just submitted this as first pass to see if you had interest in using it. That said, I do agree that using that crate probably makes more sense than the implementation I came up with.

Reply:
It does so, but this is a usual behavior of a BufReader.
The problem isn't the chunked reading, but rather the fact that it collects them into memory before streaming them to stdout.

The performant way would be to seek to the desired position in the file and stream the remaining contents directly to stdout.

It's a "read into memory and copy to stdout" vs. "stream directly to stdout". Reading into memory can be critical with the amount of data that possibly accrues

Reply:
I see what you're saying now. Yes that is true that I didn't try to address the scenario where a large number of lines are requested. It doesn't strike me as a likely use case, but understood that ideally that would be addressed as well. One thing that stopped me from even looking at that was that I wasn't sure of is whether there is currently a mechanism to stream data between client/server.  Since its message based, I would think there would need to be a mechanism for sending data in chunks from server to client which didn't jump out at me in the code but I also may have just missed it.

That said using that other crate with everything else the same, should at least prevent the extra copy of data in `read_last_lines` since it looked like it stores everything contiguously as opposed to my `VecDeque` based implementation.

Reply:
Ignore the last comment regarding streaming, of course there is a mechanism to do that. I 

> I went ahead and double-checked the `rev_buf_reader` crate. In general it looks good to me beside a few small issues.
> 
> I created a few PRs ([andre-vm/rev_buf_reader#1](https://github.com/andre-vm/rev_buf_reader/pull/1)), let's wait and see if the maintainer is still active :) We could always offer a co-maintainership.

Just glanced at this PR and it seems like its maintained based on the responses, so I'm going to go ahead and switch to it. I'll do a quick review of that crate as well just to have another set of eyes on.

Reply:
Ignore the last comment regarding streaming, of course there is a mechanism to do that. I 

> I went ahead and double-checked the `rev_buf_reader` crate. In general it looks good to me beside a few small issues.
> 
> I created a few PRs ([andre-vm/rev_buf_reader#1](https://github.com/andre-vm/rev_buf_reader/pull/1)), let's wait and see if the maintainer is still active :) We could always offer a co-maintainership.

Just glanced at this PR and it seems like its maintained based on the responses, so I'm going to go ahead and switch to it. I'll do a quick review of that crate as well just to have another set of eyes on.